packages/grafana-ui/src/components/uPlot/config.ts

```diff
@@ -0,0 +1,75 @@
+import { SelectableValue } from '@grafana/data';
+
+export enum AxisPlacement {
+  Auto = 'auto', // First axis on the left, the rest on the right
+  Top = 'top',
+  Right = 'right',
+  Bottom = 'bottom',
+  Left = 'left',
+  Hide = 'hide',
+}
+
+export function getUPlotSideFromAxis(axis: AxisPlacement) {
+  switch (axis) {
+    case AxisPlacement.Top:
+      return 0;
+    case AxisPlacement.Right:
+      return 1;
+    case AxisPlacement.Bottom:
+      return 2;
+    case AxisPlacement.Left:
+  }
+  return 3; // default everythign to the left
+}
+
+export enum PointMode {
+  Auto = 'auto', // will show points when the density is low or line is hidden
+  Always = 'always',
+  Never = 'never',
+}
+
+export enum GraphMode {
+  Line = 'line', // default
+  Bar = 'bar', // will also have a gap percent
+  Points = 'points', // use the points config
+  Staircase = 'staircase', // https://leeoniya.github.io/uPlot/demos/line-stepped.html
+  Smooth = 'smooth', // https://leeoniya.github.io/uPlot/demos/line-smoothing.html
+}
+
+export interface GraphFieldConfig {
+  mode: GraphMode;
+  lineWidth: number; // pixels
+  fillAlpha: number; // 0-1
+
+  points: PointMode;
+  pointRadius: number; // pixels
+  symbol: string; // eventually dot,star, etc
+
+  // Axis is actually unique based on the unit... not each field!
+  axisPlacement: AxisPlacement;
+  axisLabel?: string;
+  axisWidth?: number; // pixels ideally auto?
```

Comment:
yes!

Reply:
we can get rid of axisWidth

Reply:
i think we will want this if there's any hope of suppprting https://github.com/grafana/grafana/issues/4469.

my suggestion was to have a toggle between auto and explicit size.

we'll want this for gutter as well.

is there a way to retain these settings in the panel json but not expose them in the ui, eg. for advanced users?

Reply:
the gutter is set in the code, not sure we need an option for gutter.

But yes to align the y-axis on panels placed on top of each other manually assigning width could be useful. Think we can maybe have an advanced section to hide uncommon options.  

> is there a way to retain these settings in the panel json but not expose them in the ui, eg. for advanced users?

Yes, that is possible :)

Reply:
> the gutter is set in the code, not sure we need an option for gutter.

well, if the gutter becomes dynamic (auto) as we've been discussing to support edge labels, then the plotting area width becomes unpredictable and you won't be able to guarantee alignment. it's exactly the same problem as not having explicit control of the axis width on the left.